Libraries/LibGfx/JPGLoader.cpp

```diff
@@ -0,0 +1,1057 @@
+                context.huffman_stream.stream.append(last_byte);
+                continue;
+            }
+            Marker marker = 0xFF00 | current_byte;
+            if (marker == JPG_EOI)
+                return true;
+            if (marker >= JPG_RST0 && marker <= JPG_RST7) {
+                context.huffman_stream.stream.append(marker);
+                stream >> current_byte;
+                continue;
+            }
+            dbg() << stream.offset() << String::format(": Invalid marker: %x!", marker);
+            return false;
+        } else {
+            context.huffman_stream.stream.append(last_byte);
+        }
+    }
+
+    ASSERT_NOT_REACHED();
+}
+
+static bool load_jpg_impl(JPGLoadingContext& context)
+{
+    ByteBuffer buffer = ByteBuffer::wrap(context.compressed_data, context.compressed_size);
+    BufferStream stream(buffer);
+    if (!parse_header(stream, context))
+        return false;
+    if (!scan_huffman_stream(stream, context))
+        return false;
+
+    auto result = decode_huffman_stream(context);
+    if (!result.has_value()) {
+        dbg() << stream.offset() << ": Failed to decode Macroblocks!";
+        return false;
+    } else {
+        auto macroblocks = result.release_value();
+        dbg() << String::format("%i macroblocks decoded successfully :^)", macroblocks.size());
+        dequantize(context, macroblocks);
+        inverse_dct(context, macroblocks);
+        ycbcr_to_rgb(context, macroblocks);
+        compose_bitmap(context, macroblocks);
+    }
+    return true;
+}
+
+JPGImageDecoderPlugin::JPGImageDecoderPlugin(const u8* data, size_t size)
+{
+    m_context = make<JPGLoadingContext>();
+    m_context->compressed_data = data;
+    m_context->compressed_size = size;
+    m_context->huffman_stream.stream.ensure_capacity(500 * KB);
+}
+
+JPGImageDecoderPlugin::~JPGImageDecoderPlugin()
+{
+}
+
+IntSize JPGImageDecoderPlugin::size()
+{
+    if (m_context->state == JPGLoadingContext::State::Error)
+        return {};
+    if (m_context->state >= JPGLoadingContext::State::FrameDecoded)
+        return { m_context->frame.width, m_context->frame.height };
+
+    return {};
+}
+
+RefPtr<Gfx::Bitmap> JPGImageDecoderPlugin::bitmap()
+{
+    if (m_context->state == JPGLoadingContext::State::Error)
+        return nullptr;
+    if (m_context->state < JPGLoadingContext::State::BitmapDecoded) {
+        if (!load_jpg_impl(*m_context)) {
+            m_context->state = JPGLoadingContext::State::Error;
+            return nullptr;
+        }
+        m_context->state = JPGLoadingContext::State::BitmapDecoded;
+    }
+
+    return m_context->bitmap;
+}
+
+void JPGImageDecoderPlugin::set_volatile()
+{
+    if (m_context->bitmap)
+        m_context->bitmap->set_volatile();
+}
+
+bool JPGImageDecoderPlugin::set_nonvolatile()
+{
+    if (!m_context->bitmap)
+        return false;
+    return m_context->bitmap->set_nonvolatile();
+}
+
+bool JPGImageDecoderPlugin::sniff()
+{
+    return false;
+}
```

Comment:
I think that code should be functional before going in branch.. Since JPEG files have a "header" of sorts, maybe reading the first marker and checking whether it's JPG_SOI should be enough ?
(apparently, checking a six-byte "FF D8 FF" sequence is also a possibility for JPG files. Since the third and fourth byte should also be a valid marker, then it should start with "FF".)
```suggestion
{
    return m_context->compressed_size > 3
        && m_context->compressed_data[0] == 0xFF
        && m_context->compressed_data[1] == 0xD8
        && m_context->compressed_data[2] == 0xFF;
}
```